### ISSUE TITLE
False positive on "can be declared as return type hint"

### ISSUE THREAD

**Issue by iltar** [kalessil/phpinspectionsea#257]
This check is really neat! However, it's giving me quite a few false positives on interface that come from a vendor package. This is a shortened example from the Symfony `UserProviderInterface`.

```php
namespace Symfony\Component\Security\Core\User;

interface UserProviderInterface
{
    public function supportsClass($class);
}
```

```php
namespace App\Component\Security\Authentication\User

final class UserProvider implements UserProviderInterface
{
    public function supportsClass($class)
    {
        return $class === $this->class || is_subclass_of($class, $this->class);
    }
}
```
![image](https://cloud.githubusercontent.com/assets/1754678/25125266/80f2c676-242e-11e7-90fb-b49007f68bfe.png)

Yes, this check is correct. However, I cannot change the vendor code. If I were to do this, it would change the vendor code and crash on our CI because the return type declaration is incompatible (fresh composer install there without my changes).

I know this will be a difficult thing to fix, but right now, this check is dangerous to blindly follow (which people will do).

**Comment by kalessil** [comment]
Great to hear =)

What happens if you add the proposed type hint?

**Comment by rentalhost** [comment]
For what I understand, the problem here is that the `/vendor` is inspectioned. But it is inevitable. I guess that what you should do is ignore the vendor directory on _inpection code_. In general way, it could be useful if you is the vendor dev (for instance, a subpackage) and you'll like to know that you have a improvement in your code. In case where you are not the owner, just ignore that.

**Comment by kalessil** [comment]
@rentalhost : warning is in user class, the question is about re-defining the implementation signature.

@iltar, safely apply type hints for this case: https://3v4l.org/l86ir

**Comment by iltar** [comment]
> What happens if you add the proposed type hint?

Right now, it doesn't fix the interface in the vendor and it PHP will allow this (as the return type of the implementation is stricter and doesn't conflict with the empty definition).

This still means that the interfaces are not matching though, but I guess that should be fine (as long as people don't rely on the vendor changes). I just noticed that if the parent method (docblock) changes to int, it will stop telling me to add a `: bool` and work as intended.

Thanks for checking it out!

**Comment by kalessil** [comment]
For this case just introduce your own interface (inherited from the 3rd party) and fix type-hints there, less headache on long-term timeframe.

**Comment by kalessil** [comment]
By fix I mean define them implicitly.
